### ISSUE TITLE
New rule request: no-padding (no empty lines as first or last child in a block)

### ISSUE THREAD

**Issue by brunolm** [palantir/tslint#3969]
# New Rule Request: no-padding

Disallow empty lines as first or last child in a block.

```json
"no-padding": true
```

## Examples

```ts
class Foo {
          ~       // <-- no-padding
  constructor() {}
                  // <-- no-padding
}
~
```

```ts
class Foo {
  constructor() {}
                  // <-- no-padding
}
~
```

```ts
class Foo {
          ~       // <-- no-padding
  constructor() {}
}
```

```ts
const hero = {
             ~  // <-- no-padding
  id: 1,
                // <-- no-padding
}
~
```

```ts
const heroes = [
               ~   // <-- no-padding
  { id: 1 },
  { id: 2 },
                   // <-- no-padding
]
~
```

## Source draft

```ts
// tslint:disable:max-classes-per-file no-magic-numbers
import * as Lint from 'tslint';
import { getNextToken } from 'tsutils';
import * as ts from 'typescript';

export class Rule extends Lint.Rules.AbstractRule {
  public static failureMessage = 'padding space not allowed';

  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
    return this.applyWithWalker(new NoPaddingWalker(sourceFile, this.getOptions()));
  }
}

class NoPaddingWalker extends Lint.RuleWalker {
  static openPattern = /^[\[{]\n\s*\n/;
  static closePattern = /\n\s*\n\s*[\]}]$/;

  protected visitArrayLiteralExpression(node: ts.ArrayLiteralExpression): void {
    this.checkBlock(node);

    super.visitArrayLiteralExpression(node);
  }

  protected visitBlock(node: ts.Block): void {
    this.checkBlock(node);

    super.visitBlock(node);
  }

  protected visitClassDeclaration(node: ts.ClassDeclaration): void {
    const children = node.getChildren();

    children.reduce((current, next) => {
      if (!current) {
        return next;
      }

      if (current.kind === ts.SyntaxKind.FirstPunctuation) {
        const nextToken = getNextToken(next);

        const text = current.getText();
        const nextFullText = next.getFullText();
        const afterNextText = nextToken.getFullText();

        this.checkBlock(next, `${text}${nextFullText}${afterNextText}`);
      }

      return next;
    });

    super.visitClassDeclaration(node);
  }

  protected visitObjectLiteralExpression(node: ts.ObjectLiteralExpression): void {
    this.checkBlock(node);

    super.visitObjectLiteralExpression(node);
  }

  private checkBlock(node: ts.Node, fullText?: string): void {
    const text = fullText || node.getText();

    const isOpenInvalid = NoPaddingWalker.openPattern.test(text);
    const isCloseInvalid = NoPaddingWalker.closePattern.test(text);

    if (isOpenInvalid) {
      const start = fullText ? node.getFullStart() - 1 : node.getStart();

      this.addFailureAt(start, 3, Rule.failureMessage);
    }

    if (isCloseInvalid) {
      const nodeEnd = node.getEnd();
      const start = fullText ? getNextToken(node).getStart() : nodeEnd - 1;

      this.addFailureAt(start, 1, Rule.failureMessage);
    }
  }
}
```

## Preview

![no-padding-tslint](https://user-images.githubusercontent.com/706078/41504344-1d3aaa44-71c2-11e8-854b-d964f4d5cc38.gif)

**Comment by giladgray** [comment]
@brunolm does the `whitespace` rule capture this? if not, that's the place for these rules.

https://palantir.github.io/tslint/rules/whitespace/

also, i use prettier with tslint which definitely fixes whitespace like this.

**Comment by brunolm** [comment]
hmm, `whitespace` doesn't seem to do that exactly.

But yeah, using prettier is an option, I even researched how to do it before a commit, so there's no way to forget it.

```
npm i -D lint-staged husky@^1.0.0-rc.9 prettier
```

package.json
```
"husky": {
   "hooks": {
     "pre-commit": "lint-staged"
   }
 },
 "lint-staged": {
   "*.{ts,tsx}": [
     "prettier --write"
   ]
 },
```

**Comment by michaelcm** [comment]
Just checking if there is a way to do this with `tslint` yet, or if the best option is still `prettier`?

**Comment by adidahiya** [comment]
the best option for all code formatting functionality, IMO, is to use prettier. tslint is de-prioritizing formatting-related rules: #3592

**Comment by JimLynchCodes** [comment]
Personally, I would like a tslint rule that enforces one empty line after a class / function signature... is this possible?

**Comment by adidahiya** [comment]
@JimLynchCodes see #4534
